src/dp_abstraction/aca_goal_state_handler.cpp

```diff
@@ -140,37 +149,44 @@ int Aca_Goal_State_Handler::update_port_states(GoalState &parsed_struct,
   int rc;
   int overall_rc = EXIT_SUCCESS;
   int count = 1;
-
+  GoalState* gs_ptr = &parsed_struct;
+  GoalStateOperationReply* reply_ptr = &gsOperationReply;
+  marl::WaitGroup wait_group(parsed_struct.port_states_size());
   for (int i = 0; i < parsed_struct.port_states_size(); i++) {
     ACA_LOG_DEBUG("=====>parsing port states #%d\n", i);
 
     PortState current_PortState = parsed_struct.port_states(i);
-
-    workitem_future.push_back(std::async(
-            std::launch::async, &Aca_Goal_State_Handler::update_port_state_workitem, this,
-            current_PortState, std::ref(parsed_struct), std::ref(gsOperationReply)));
-    if (count % resource_state_processing_batch_size == 0){
-      for (int i = 0; i < workitem_future.size(); i++) {
-        rc = workitem_future[i].get();
-        if (rc != EXIT_SUCCESS)
-          overall_rc = rc;
-      }
-      workitem_future.clear();
-      count = 1;
-    } else {
-      count ++;
-    }
+    
+    marl::schedule([=] {
+      defer(wait_group.done());
+      // defer(neighbor_wait_group.done());
+      update_port_state_workitem(current_PortState, *gs_ptr, *reply_ptr);
+    });
+    // workitem_future.push_back(std::async(
+    //         std::launch::async, &Aca_Goal_State_Handler::update_port_state_workitem, this,
+    //         current_PortState, std::ref(parsed_struct), std::ref(gsOperationReply)));
+    // if (count % resource_state_processing_batch_size == 0){
+    //   for (int i = 0; i < workitem_future.size(); i++) {
+    //     rc = workitem_future[i].get();
+    //     if (rc != EXIT_SUCCESS)
+    //       overall_rc = rc;
+    //   }
+    //   workitem_future.clear();
+    //   count = 1;
+    // } else {
+    //   count ++;
+    // }
     // keeping below just in case if we want to call it serially
     // rc = update_port_state_workitem(current_PortState, parsed_struct, gsOperationReply);
     // if (rc != EXIT_SUCCESS)
     //   overall_rc = rc;
   } // for (int i = 0; i < parsed_struct.port_states_size(); i++)
-
-  for (int i = 0; i < workitem_future.size(); i++) {
-    rc = workitem_future[i].get();
-    if (rc != EXIT_SUCCESS)
-      overall_rc = rc;
-  } // for (int i = 0; i < parsed_struct.port_states_size(); i++)
+  wait_group.wait();
```

Comment:
if  wait_group   is not stable, we do not need it in master ACA

Reply:
We shall keep the waitgroup as we need to wait for all states to be finished until going into the next step.